langchain/chains/combine_documents/map_reduce.py

```diff
@@ -49,14 +55,79 @@ def get_default_document_variable_name(cls, values: Dict) -> Dict:
                 )
         return values
 
-    def combine_docs(self, docs: List[Document], **kwargs: Any) -> str:
-        """Combine by mapping first chain over all, then stuffing into final chain."""
+    def compute_tokens(self, results: List, docs: List[Document], **kwargs: Any) -> int:
+        """Checks to see if, after combining a set of documents into one text, if it is too long."""
+        question_result_key = self.llm_chain.output_key
+        result_docs = [
+            Document(page_content=r[question_result_key], metadata=docs[i].metadata)
+            # This uses metadata from the docs, and the textual results from `results`
+            for i, r in enumerate(results)
+        ]
+        doc_dicts = []
+        for doc in result_docs:
+            base_info = {"page_content": doc.page_content}
+            base_info.update(doc.metadata)
+            document_info = {
+                k: base_info[k] for k in self.document_prompt.input_variables
+                # TODO: how does this work with plain QA, as opposed to this which is designed to work with
+                # QA with sources right now?
+            }
+            doc_dicts.append(document_info)
+        # Format each document according to the prompt
+        doc_strings = [self.document_prompt.format(**doc) for doc in doc_dicts]
+        combined_texts = "\n\n".join(doc_strings)
+
+        # tokenize the text and compute num tokens
+        tokenizer = GPT2TokenizerFast.from_pretrained('gpt2')
+        tokenized_text = tokenizer.tokenize(combined_texts)
+        num_tokens = len(tokenized_text)
+        return num_tokens
+
+    def combine_docs(self, docs: List[Document],
+                     token_max=3000,
+                     **kwargs: Any) -> str:
+        """
+        Combine by mapping first chain over all, then
+        if NUM OF TOKENS OF combined_texts is too large for context window,
+        combine some text and run self.llm_chain.apply again,
+        keep doing this in a while loop until the num of tokens in the combined text fits, then
+        stuff it into a final chain once the NUM OF TOKENS OF combined_texts is within the limit.
+        """
+
         results = self.llm_chain.apply(
+            # FYI - this is parallelized and so it is fast.
             [{**{self.document_variable_name: d.page_content}, **kwargs} for d in docs]
         )
         question_result_key = self.llm_chain.output_key
+
+        num_tokens = self.compute_tokens(results, docs, **kwargs)
+        print(num_tokens)
+
+        # and if NUM OF TOKENS OF combined_texts is too large for context window,
+        # then combine some texts and keep running self.llm_chain.apply again until it fits.
+        while num_tokens > token_max:
+            # docs is a List[Document] where each Document has .page_content w/ the text
+            alld = []
+            for i in range(len(results)):
+                if (i % 2) != 0:
```

Comment:
2 is a guess, right? could we try to do something where we combine as many as is possbile?

Reply:
Yeah, but how would we compute that without trying all possible permutations?

Reply:
can we do it in a for loop? like loop through it and create a list of list dynamically (potentially with different number documents? that shouldnt matter right?)

Reply:
what is the rough pseudo-code you have in mind here?

Reply:
(Recall that you can't evaluate the length of the result until you run another LLM Q/A on each component, at least as far as I am thinking about it)